docs/blog/2023-07-15-incremental.md

```diff
@@ -0,0 +1,226 @@
+two versions of a program.  Important activities of programmers, such
+as [code reviews](https://en.wikipedia.org/wiki/Code_review), often
+only look at changes in programs.
+
+There are specific tools, like
+[diff](https://en.wikipedia.org/wiki/Diff) which compute changes
+between text files.  Here is an example from the above Wikipedia page:
+
+```console
+$ diff original new
+0a1,6
+> This is an important
+> notice! It should
+> therefore be located at
+> the beginning of this
+> document!
+>
+11,15d16
+< This paragraph contains
+< text that is outdated.
+< It will be deleted in the
+< near future.
+17c18
+< check this dokument. On
+---
+> check this document. On
+```
+
+Here both `original` and `new` are text files.  The `diff` tool
+operates on *lines*, and shows how lines in one file have to be
+modified to generate the second file.  The tool `diff` describes the
+difference as a set of *additions*, (abbreviated `a`), *deletions*
+(`d`) and changes (`c`).  The above output shows that to change file
+`original` into `new` one has to add 6 lines at starting from line 0,
+delete lines 11 to 15, and then replace line 17.  The `<` sign points
+to lines in the left file (`original`), while the `>` sign points to
+lines in the right file.
+
+Notice that the difference between two files can be described in
+multiple ways: another way would be to delete everything in the first
+file and then add everything in the second file.  The `diff` tool
+tries to find a *small* set of changes, modifying as few lines as
+possible.  Also, unlike Word, `diff` does not show character changes,
+only full line changes.
+
+Also, notice that `diff` changes are described as being composed of
+multiple simpler changes: the example contains one insertion, one
+deletion, and one modification.
+
+`diff` can compare not only two files but it can also compare multiple
+files at once.  Entire files can be inserted, deleted, renamed, or
+changed individually.  Another tool,
+[patch](https://en.wikipedia.org/wiki/Patch_(Unix)) can use such a
+description to convert a set of old file versions into a set of newer
+ones.
+
+> Changes can be hierarchical: changes to complex objects can be
+  described from changes to simpler ones
+
+## Data and changes
+
+Notice that we are dealing with two different kinds of structures: the
+changed objects (e.g., files), and the changes between objects
+(deltas, or diffs).  (One can store a change in a text file, but its
+really a different object.)
+
+Let's call these two data structures *D* for data, and *Δ* for
+changes.  First we have a *diff8 operation *D - D → Δ*: where use
+minus for the *diff*.  The diff between two data values is a change.
+
+We also have an inverse operation: *D + Δ → D*, which changes some
+data by *applying* a change.  There is also a notion of an *empty
+change*, which is the difference between two identical data values.
+
+## Changing numbers
+
+In some circumstances the data and the changes can look similar.
```

Comment:
I expected that you'd at some point say that this is true for databases too.

Reply:
I reserved a blog post about zsets to talk about this in more details.
Planned for later in the series.

Reply:
Fair enough. A teaser would be nice though; otherwise this feels like a loose end.

Reply:
I have added the following paragraph (next commit):

For this case the structures D and Δ are identical: you can represent both the data and the changes as numbers (but you must include negative numbers if you want to represent all possible changes). Empty data and empty changes are both the number zero. This is very convenient. We will show up in a future blog post that we can take advantage of this nice property even when we don't handle just numbers.